### ISSUE TITLE
OpenSSL 1.1.0 not compatible with specific web server

### ISSUE THREAD

**Issue by cschramm** [openssl/openssl#2669]
I'm experiencing strange behavior with HTTPS connections from my Debian system to several servers in applications that use OpenSSL 1.1.0, including Python.

I can reproduce the issue with e.g. `echo 'GET / HTTP/1.0\nHost: pintan.santanderbank.de\n\n' | openssl s_client -quiet -connect pintan.santanderbank.de:443`. With OpenSSL 1.0.2 this just displays the HTTP response followed by `read:errno=0` and terminates as expected. With OpenSSL 1.1.0 the connection does not get closed and the server sends the response over and over again in an endless loop. I didn't see the encrypted alert I'd expect the server to send in Wireshark.

I have no idea what I'm actually experiencing there and if it's really an issue with OpenSSL (probably in connection with a specific server-side library). The versions I tried are 1.0.2k and 1.1.0d Debian packages.

**Comment by sebastianas** [comment]
I doubt this a problem in openssl. You have the same behaviour (endless response from the server) with gnutls-cli as well.

**Comment by sebastianas** [comment]
A bisect shows that c43a55407dccc6902058184d7dd0bd111fe6a61e is responsible for the endless loop.

**Comment by mattcaswell** [comment]
I suspect this is a server side issue. Issuing the above command against `openssl s_server -www` sees s_client behaving as you would expect, i.e. it sends the GET request and receives a response back from the server which then closes the connection.

In your case the server is never closing the connection. With the options that you are using s_client stays open until the server ends the connection. For some reason this server never does.

The commit @sebastianas points to switches on Encrypt-Then-Mac support. This is a newer feature that not all servers support. I wonder if you are getting directed to a different set of web servers in the server farm when using this which are behaving differently??

**Comment by cschramm** [comment]
I'm sure it's related to a specific (Java) server software. I discovered the behavior with a couple of different web sites but only in case chunked transfer encoding gets used at some point because the chunks are not complete (the 0-size chunk does not appear) before the HTTP status line gets repeated and I get an IncompleteRead exception in Python. I could not find out which web server software is used. BuiltWith reports the J2EE framework for the two samples I have in mind (the second one being www.banking.co.at).

The trigger might be described in #2691 although the sample addresses in the NSS bug report show a very different behavior (i.e. issues at the TLS layer, not in the payload) and I doubt that the services I'm having trouble with are using Tomcat. Not sending the empty EtM (and the EMS) extension last might help, though.

**Comment by mattcaswell** [comment]
There does not seem any evidence that this is an OpenSSL issue. This appears to be a server side problem. Without further information I am closing this issue.

**Comment by cschramm** [comment]
So OpenSSL 1.1 is not meant to be compatible to existing real world SSL applications? Ugh, that's really disappointing... 😕

**Comment by mattcaswell** [comment]
Of course it is. All I'm saying is that there is no evidence that the problem is on OpenSSL's side. AFAICT OpenSSL is doing exactly the right thing. It creates the connection, and awaits for the server to close it. The fact that the server doesn't close it appears to be server side issue. There seems very little OpenSSL can do about that.

**Comment by cschramm** [comment]
I'm absolutely sure it is a server side issue, yes, but OpenSSL could avoid triggering it. I really hope you'll give not putting empty extensions at the end of the Client Hello a try, see #2691 and the NSS bug referenced there.

**Comment by mattcaswell** [comment]
> I'm absolutely sure it is a server side issue, yes, but OpenSSL could avoid triggering it

Possibly. Possibly not. It depends on what the issue is. But without a diagnosis on the server side that's going to be very difficult/impossible for us to do.

> I really hope you'll give not putting empty extensions at the end of the Client Hello a try, see #2691 and the NSS bug referenced there.

That's a different issue. It seems unlikely that this is the same problem. That issue will be considered on its own merits independently.

**Comment by cschramm** [comment]
@sebastianas: How do you get the behavior with gnutls-cli? I cannot replicate it with either version 3.3 or 3.5.

**Comment by sebastianas** [comment]
It should have been 3.5.8-3 but I can't reproduce it right now.
gnutls 3.3.8-6+deb8u5 gives me no output at all. It says:
```
Processed 162 CA certificate(s).
Resolving 'pintan.santanderbank.de'...
…
- Handshake was completed

- Simple Client Mode:

"shell"
```
As for openssl: if you strip the -quiet part from openssl then it "works" so this looks very timing related.

**Comment by sebastianas** [comment]
btw, with the master branch you can do something like
`SSL_CTX_set_options(ctx, SSL_OP_NO_ENCRYPT_THEN_MAC);
`
and disable EtM. I haven't found anything like that in the 1.1.0 branch and a backport doesn't seem that simple. However, EtM won't be use for AEAD ciphers so something like
` -cipher 'EECDH+AESGCM:EDH+AESGCM'
`
added to the openssl command does not trigger the bug on the server side as well.

**Comment by cschramm** [comment]
@sebastianas: Thanks a bunch for those hints! So this will actually be resolved in 1.1.1 with `SSL_OP_NO_ENCRYPT_THEN_MAC`. And even better, using cipher list `AESGCM:AESCCM:CHACHA20:!PSK:!ADH` renders a decent workaround for me to get EtM disabled as you suggested (as long as the servers provide any of them). 😃
